### PR TITLE
Language Switcher Widget for Elementor

### PR DESCRIPTION
**Pull request by amElnagdy**
[I've been asked](https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-6553) to create a language switcher widget for Elementor.
Thanks to @strategio's help, the widget is almost ready.

Note for me, more selectors still needed to cover various options in WPML->Languages -> custom language switchers

### CODE REVIEW

File: src/language-switcher/class-wpml-elementor-language-switcher-factory.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
```
Comment (strategio) [inline src/language-switcher/class-wpml-elementor-language-switcher-factory.php]: @amElnagdy, now that we use namespaces, you can use smaller file names => `factory.php` is enough :)

**Review by strategio — COMMENTED** [review]
@amElnagdy, it sounds good. I have no very deep knowledge of the Elementor API to build custom widgets but what I see makes sense to me. It would be good if we could have some unit tests. Is this something you are planning to do?

File: src/class-wpml-elementor-integration-factory.php [hunk 2]
```diff
@@ -20,7 +20,8 @@ public function create() {
 				'WPML_Elementor_Adjust_Global_Widget_ID_Factory',
 				'WPML_PB_Elementor_Handle_Custom_Fields_Factory',
 				'WPML_Elementor_Media_Hooks_Factory',
-				'WPML_Elementor_WooCommerce_Hooks_Factory'
+				'WPML_Elementor_WooCommerce_Hooks_Factory',
+				'\WPML\PB\Elementor\LanguageSwitcher\Factory'
```
Comment (brucepearson) [inline src/class-wpml-elementor-integration-factory.php]: @amElnagdy We no longer need factory classes for simple hooks now. You can change this to `'\WPML\PB\Elementor\LanguageSwitcher`. I'll add comments in other places about what else to change.

File: src/language-switcher/class-wpml-elementor-language-switcher-factory.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+namespace WPML\PB\Elementor\LanguageSwitcher;
+
+class Factory implements \IWPML_Backend_Action_Loader, \IWPML_Frontend_Action_Loader {
```
Comment (brucepearson) [inline src/language-switcher/class-wpml-elementor-language-switcher-factory.php]: @amElnagdy This factory wont be needed

File: src/language-switcher/class-wpml-elementor-language-switcher.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace WPML\PB\Elementor\LanguageSwitcher;
+
+class LanguageSwitcher implements \IWPML_Action {
```
Comment (brucepearson) [inline src/language-switcher/class-wpml-elementor-language-switcher.php]: @amElnagdy You can change this to implement `\IWPML_Backend_Action` and `\IWPML_Frontend_Action` and you wont need the factory.

**Review by brucepearson — CHANGES_REQUESTED** [review]
@amElnagdy I've made some suggested changes. Can we have unit tests for all this?

File: src/language-switcher/class-wpml-elementor-language-switcher-factory.php (hunk 1 above)
Comment (amElnagdy) [reply src/language-switcher/class-wpml-elementor-language-switcher-factory.php]: Perfect, thanks @strategio!

As I have you here, to make this whole thing works properly the custom language switcher option in WPML -> Languages has to be enabled and the Dropdown option is selected. 
(As we already using wpml language switcher action)
So, should we enable it programmatically (is it possible?) or ask the clients to do?
What do you think?

More details: I can handle the different options in the custom language switcher (dropdown, dropdown click, horizontal list, etc..) by adding more selectors to the Elementor API but the optimal thing would just have a dropdown option as other options are easily passed to the action.

Comment (strategio) [reply src/language-switcher/class-wpml-elementor-language-switcher-factory.php]: @amElnagdy, we should enable the custom language switcher programmatically. Please create a ticket in Core and assign it to me. I will provide some method to do this.

Comment (amElnagdy) [reply src/language-switcher/class-wpml-elementor-language-switcher-factory.php]: Created - https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-6669

File: src/language-switcher/language-switcher.php [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace WPML\PB\Elementor\LanguageSwitcher;
 
-class LanguageSwitcher implements \IWPML_Action {
+class LanguageSwitcher implements \IWPML_Backend_Action, \IWPML_Frontend_Action {
```
Comment (amElnagdy) [inline src/language-switcher/language-switcher.php]: @brucepearson this is not loading the class, I'm getting a bunch of errors. Would you like to check? debug.log https://pastebin.com/myNeusth

Comment (brucepearson) [reply src/language-switcher/language-switcher.php]: @amElnagdy 
This is failing - class_implements(): Class \WPML\PB\Elementor\LanguageSwitcher\LanguageSwitcher
Did you run composer install to update autoloading?

Comment (amElnagdy) [reply src/language-switcher/language-switcher.php]: Not sure if I did. Currently, it's working fine. Thanks @brucepearson

**Comment by andreasciamanna** [comment]
@amElnagdy can you please update the branch?

**Comment by strategio** [comment]
I have rebased the branch and added some tests. Unfortunately, it's almost impossible to write tests because our widget is strongly coupled with Elementor's code (multiple inheritance) and Elementor's
code is strongly coupled (massive use of static methods and properties + singleton pattern).

I managed to cover the main `LanguageSwitcher` class by including the Elementor plugin from GitHub as a composer dependency. But I could not go further, covering the `Widget` class is impossible in pure unit, it would require integration tests.

@andreasciamanna, @brucepearson, @amElnagdy

**Comment by andreasciamanna** [comment]
@amElnagdy or @strategio please add a reference to the ticket in the main commit message and in this PR.

Also please rearrange all the commits because at the moment it's not clear which one is a fixup and which one is a commit that must stay as-is.

Once this is done, make sure the ticket contains all the steps to test the widget, so we may add end-to-end tests.

**Review by andreasciamanna — DISMISSED** [review]
It seems ok, but it's also a quite complex code. See my other comment in the PR.

**Comment by strategio** [comment]
@amElnagdy, I'll let you squash/rebase the commits as you are the holder of this story.

File: src/language-switcher/widget/widget.php [hunk 6]
```diff
@@ -0,0 +1,337 @@
+<?php
+namespace WPML\PB\Elementor\LanguageSwitcher;
+
+use Elementor\Controls_Manager;
+use Elementor\Group_Control_Typography;
+use Elementor\Scheme_Color;
+use Elementor\Widget_Base;
+
+class Widget extends Widget_Base {
```
Comment (brucepearson) [inline src/language-switcher/widget/widget.php]: @strategio If you really want to test the code properly then I think you need to remove all the business logic out of this class and use an adaptor pattern. The should make this `Widget` very simple so it doesn't have any logic that you need to test with unit tests. Then you can fully test the adaptor.

```php
__construct( Adaptor $adaptor ) {
   parent::__construct();
    $this->adaptor = $adaptor;
    $adaptor->target( $this );
}

public function get_name() {
    return $this->adaptor->getName();
}

...
...

protected function _register_controls() {
    $this->adaptor->registerControls();
}

```

Then in the Adaptor you call `$this->target->start_controls_section( .... )`, etc, which you can then test to confirm the target is called with the correct data.

Comment (strategio) [reply src/language-switcher/widget/widget.php]: Thanks for the suggestion @brucepearson. This is something I was thinking but I was unsure if all the methods used in `_register_controls` and `render` were public (I am talking of `get_settings_for_display`, `add_render_attribute`, `add_control`, ...). It looks like everything could be accessed by the adaptor. I will try to implement this!

**Comment by strategio** [comment]
@brucepearson, I managed to implement the adaptor pattern but it was not so easy...

Some comments:
1. I decided to keep the elementor plugin as a dev package because we are
calling a lot of class constants which we would have to stub. I also
wanted to make sure that the methods we are using in the adaptor are
actually public (this is really not obvious).
2. One of the methods in the widget base class is
marked as "final" which does not allow mocking. I decided to remove this
when the package is installed in the dev dependencies (see script in composer.json).
3. Our `Widget` class can be instantiated by Elementor and so we can't
change the first 2 arguments in the constructor and we need to allow creating the
adaptor inside the constructor.

**Comment by strategio** [comment]
The tests are failing in Scrutinizer because some Elementor classes could not be loaded. I can't find out the problem (did several blind tests), but things are working smoothly locally. I don't want to invest too much time on this since we are planning to move the repo to GitLab.

**Review by brucepearson — DISMISSED** [review]
@strategio Nice work. +1

**Comment by strategio** [comment]
I squashed all the commits because after the transposition into the "adatpor" pattern, it does not make sense do have different commits.

**Pull request merged**